lib/index.js

```diff
@@ -109,93 +112,127 @@ internals.Connection.prototype.validateSegmentName = function (name) {
+        return callback(new Error('Connection not started'));
     }
 
     const segment = this.cache[key.segment];
+
     if (!segment) {
-        return null;
+        return callback(null, null);
     }
 
     const envelope = segment[key.id];
 
     if (!envelope) {
-        return null;
+        return callback(null, null);
     }
 
-    let item = null;
+    if (Date.now() > envelope.expireTime) {
+        return this.drop(key, () => {
+
+            callback(null, null);
+        });
+    }
+
+    let value = null;
+
     if (Buffer.isBuffer(envelope.item)) {
-        item = envelope.item;
+        value = envelope.item;
     }
     else {
-        item = internals.parseJSON(envelope.item);
+        value = internals.parseJSON(envelope.item);
+
+        if (value instanceof Error) {
+            return callback(new Error('Bad value content'));
+        }
     }
 
     const result = {
-        item,
+        item: value,
         stored: envelope.stored,
         ttl: envelope.ttl
     };
 
-    return result;
+    return callback(null, result);
 };
 
 
-internals.Connection.prototype.set = async function (key, value, ttl) {
+internals.Connection.prototype.set = function (key, value, ttl, callback) {
+
+    callback = Hoek.nextTick(callback);
 
     if (!this.cache) {
-        throw new Error('Connection not started');
+        return callback(new Error('Connection not started'));
     }
 
     if (ttl > 2147483647) {                                                         // Math.pow(2, 31)
-        throw new Error('Invalid ttl (greater than 2147483647)');
+        return callback(new Error('Invalid ttl (greater than 2147483647)'));
+    }
+
+    let envelope = null;
+    try {
+        envelope = new internals.MemoryCacheEntry(key, value, ttl, this.settings.allowMixedContent);
+    }
+    catch (err) {
+        return callback(err);
     }
 
-    const envelope = new internals.MemoryCacheEntry(key, value, ttl, this.settings.allowMixedContent);
     this.cache[key.segment] = this.cache[key.segment] || new internals.MemoryCacheSegment();
     const segment = this.cache[key.segment];
     const cachedItem = segment[key.id];
 
-    if (cachedItem && cachedItem.timeoutId) {
-        BigTime.clearTimeout(cachedItem.timeoutId);
+    if (cachedItem && Date.now() < cachedItem.expireTime) {
         this.byteSize -= cachedItem.byteSize;                   // If the item existed, decrement the byteSize as the value could be different
     }
 
-    if (this.settings.maxByteSize &&
-        (this.byteSize + envelope.byteSize > this.settings.maxByteSize)) {
-
-        throw new Error('Cache size limit reached');
+    if (this.settings.maxByteSize) {
+        if (this.byteSize + envelope.byteSize > this.settings.maxByteSize) {
+            this.flushExpiredCacheItems();
+        }
+        if (this.byteSize + envelope.byteSize > this.settings.maxByteSize) {
+            return callback(new Error('Cache size limit reached'));
```

Comment:
Perhaps a bit more explicit error message here. `Unable to store cache entry, cache size limit reached`?